src/linear_algebra/pi_tensor_product.lean

```diff
@@ -442,6 +444,63 @@ def pempty_equiv : ⨂[R] i : pempty, M ≃ₗ[R] R :=
   map_add' := linear_map.map_add _,
   map_smul' := linear_map.map_smul _, }
 
+section tmul
+
+/-- Collapse a `tensor_product` of `pi_tensor_product`s. -/
+private def tmul : (⨂[R] i : ι, M) ⊗[R] (⨂[R] i : ι₂, M) →ₗ[R] ⨂[R] i : ι ⊕ ι₂, M :=
+tensor_product.lift
+  { to_fun := λ a, pi_tensor_product.lift $ pi_tensor_product.lift
+      (multilinear_map.curry_sum_equiv R _ _ M _ (tprod R)) a,
+    map_add' := λ a b, by simp only [linear_equiv.map_add, linear_map.map_add],
+    map_smul' := λ r a, by simp only [linear_equiv.map_smul, linear_map.map_smul], }
+
+private lemma tmul_apply (a : ι → M) (b : ι₂ → M) :
+  tmul ((⨂ₜ[R] i, a i) ⊗ₜ[R] (⨂ₜ[R] i, b i)) = ⨂ₜ[R] i, sum.elim a b i :=
+begin
+  erw [tensor_product.lift.tmul, pi_tensor_product.lift.tprod, pi_tensor_product.lift.tprod],
+  refl
+end
+
+/-- Expand `pi_tensor_product` into a `tensor_product` of two factors. -/
+private def tmul_symm : ⨂[R] i : ι ⊕ ι₂, M →ₗ[R] (⨂[R] i : ι, M) ⊗[R] (⨂[R] i : ι₂, M) :=
+pi_tensor_product.lift begin
+  apply multilinear_map.dom_coprod,
+  exact tprod R,
+  exact tprod R,
+end
+
+private lemma tmul_symm_apply (a : ι ⊕ ι₂ → M) :
+  tmul_symm (⨂ₜ[R] i, a i) = (⨂ₜ[R] i, a (sum.inl i)) ⊗ₜ[R] (⨂ₜ[R] i, a (sum.inr i)) :=
+pi_tensor_product.lift.tprod _
+
+variables (R M)
+
+/-- Equivalence between a `tensor_product` of `pi_tensor_product`s and a single
+`pi_tensor_product` indexed by a `sum` type.
+
+For simplicity, this is defined only for homogeneously- (rather than dependently-) typed components.
+-/
+def tmul_equiv : (⨂[R] i : ι, M) ⊗[R] (⨂[R] i : ι₂, M) ≃ₗ[R] ⨂[R] i : ι ⊕ ι₂, M :=
+linear_equiv.of_linear tmul tmul_symm
+  (by {
+    ext x,
+    show tmul (tmul_symm (tprod R x)) = tprod R x,
+    simp only [tmul_symm_apply, tmul_apply, sum.elim_comp_inl_inr], })
+  (by {
+    ext x y,
+    show tmul_symm (tmul (tprod R x ⊗ₜ[R] tprod R y)) = tprod R x ⊗ₜ[R] tprod R y,
+    simp only [tmul_apply, tmul_symm_apply, sum.elim_inl, sum.elim_inr], })
```

Comment:
Why not just:
```suggestion
linear_equiv.of_linear tmul tmul_symm
  (by { ext x, simp [tmul_symm_apply, tmul_apply, sum.elim_comp_inl_inr] })
  (by { ext x y, simp [tmul_apply, tmul_symm_apply, sum.elim_inl, sum.elim_inr] })
```

Reply:
I think that was substantially slower

Reply:
`ext` unfortunately leaves a rather messy goal, it seemed easiest to just use show to put it in the most readable form.

Reply:
```suggestion
linear_equiv.of_linear tmul tmul_symm
  (by { ext x,
        show tmul (tmul_symm (tprod R x)) = tprod R x, -- Speed up the call to `simp`.
        simp only [tmul_symm_apply, tmul_apply, sum.elim_comp_inl_inr], })
  (by { ext x y,
        show tmul_symm (tmul (tprod R x ⊗ₜ[R] tprod R y)) = tprod R x ⊗ₜ[R] tprod R y,
        simp only [tmul_apply, tmul_symm_apply, sum.elim_inl, sum.elim_inr], })
```